Refactor: integrate NetworkPolicyGateway into LDAP session management

Misha-Shvets opened the pull request:
- Updated `LDAPSession` to accept `NetworkPolicyGateway` for session creation.
- Refactored `get_session` methods in `HTTPProvider` and `LDAPServerProvider` to utilize the new gateway.
- Replaced direct session calls with `NetworkPolicyGateway` in various components, including `AuthManager`, `MFAManager`, and request handling.
- Removed deprecated network policy functions and streamlined policy checks using the new gateway methods.
- Updated tests to reflect changes in session management and policy validation.

Commit:
Refactor: integrate NetworkPolicyGateway into LDAP session management

- Updated `LDAPSession` to accept `NetworkPolicyGateway` for session creation.
- Refactored `get_session` methods in `HTTPProvider` and `LDAPServerProvider` to utilize the new gateway.
- Replaced direct session calls with `NetworkPolicyGateway` in various components, including `AuthManager`, `MFAManager`, and request handling.
- Removed deprecated network policy functions and streamlined policy checks using the new gateway methods.
- Updated tests to reflect changes in session management and policy validation.

Commit:
Refactor: reorganize network policy imports and remove unused constants

- Consolidated imports from `ldap_protocol.policies.network` for better readability.
- Removed the `constans.py` file as it was no longer needed.
- Updated references to `ProtocolType` and other network policy components across multiple files to reflect the new import structure.
- Ensured all affected files and tests are aligned with the changes in import paths.

Commit:
Refactor: enhance network policy imports and add exception handling

- Added new exception classes `LastActivePolicyError` and `NetworkPolicyAlreadyExistsError` to improve error handling in network policy operations.
- Reorganized import statements for better clarity and structure in `use_cases.py`.

Commit:
Remove: delete unused network_policy.py file

- Eliminated the `network_policy.py` file as it is no longer needed in the project structure, streamlining the codebase.

Mastermind-U commented on app/ldap_protocol/policies/network/constants.py:
```diff
@@ -0,0 +1,15 @@
+from enum import StrEnum
+
+
+class ProtocolType(StrEnum):
```
app/enums

`NetworkProtocol`
без `IS_`

Mastermind-U commented on app/ldap_protocol/policies/network/use_cases.py:
```diff
@@ -189,6 +190,64 @@ async def swap_priorities(self, id1: int, id2: int) -> SwapPrioritiesDTO:
             priority2=policy2.priority,
         )
 
+    async def get_user_http_policy(
```
Это нарушение ISP, сделай отдельный гейтвей и usecase, назови NetworkPolicyValidator[UseCase/Gateway]
Для гейтвея сделай протокол, см как сделано в rootDSE

Mastermind-U commented on app/ldap_protocol/policies/network/use_cases.py:
```diff
@@ -189,6 +190,64 @@ async def swap_priorities(self, id1: int, id2: int) -> SwapPrioritiesDTO:
+        ip: IPv4Address | IPv6Address,
+        user: User,
+    ) -> NetworkPolicy | None:
+        """Get user HTTP policy."""
```
Докстринги в методы и классы можно теперь не ставить, я отключил это правило, они тут не нужны в однострочных методах

Naksen committed:
fix: update push branch from main to dev in build workflow (#866)

Commit:
Refactor: introduce ProtocolType enum and update network policy validation

- Added a new `ProtocolType` enum to standardize protocol field references across the codebase.
- Replaced instances of the deprecated `NetworkPolicyUseCase` with `NetworkPolicyValidatorProtocol` for improved validation handling.
- Updated various components, including `AuthManager`, `MFAManager`, and request handling, to utilize the new validation structure.
- Adjusted tests to align with the changes in protocol type references and validation methods.

TheMihMih commented on app/ldap_protocol/auth/auth_manager.py:
```diff
@@ -32,9 +32,8 @@
-from ldap_protocol.policies.network_policy import (
-    check_mfa_group,
-    get_user_network_policy,
+from ldap_protocol.policies.network.validator_protocol import (
```
Ты вроде использовал импорты из пакета, а тут прям из файла. Делай в одном стиле 
Тем более в init ты его добавил

Misha-Shvets replied:
бебе

TheMihMih commented on app/ldap_protocol/ldap_requests/bind.py:
```diff
@@ -173,11 +158,12 @@ async def handle(
             return
-
-        if not await self.is_user_group_valid(
+        policy = getattr(ctx.ldap_session, "policy", None)
```
А зачем getattr? Не лучше ли будет через точку это делать?

Misha-Shvets replied:
нет, т.к. атрибута может не быть

TheMihMih commented on app/ldap_protocol/ldap_requests/bind.py:
```diff
@@ -173,11 +158,12 @@ async def handle(
-        if not await self.is_user_group_valid(
+        policy = getattr(ctx.ldap_session, "policy", None)
+        if (
+            policy is not None
```
А что такое policy? Это может быть 0 или ""?  Кажется что это инстанс какого-то класса, может можно is None убрать?

Misha-Shvets replied:
а гетатр шутка какая-то?

TheMihMih commented on app/ldap_protocol/ldap_requests/bind.py:
```diff
@@ -173,11 +158,12 @@ async def handle(
             user,
-            ctx.ldap_session,
-            ctx.session,
+            ctx.ldap_session.policy,
```
У тебя уже есть policy

TheMihMih commented on app/ldap_protocol/ldap_requests/bind.py:
```diff
@@ -192,14 +178,18 @@ async def handle(
-        if (
-            (policy := getattr(ctx.ldap_session, "policy", None))
-            and policy.mfa_status in (MFAFlags.ENABLED, MFAFlags.WHITELIST)
+        if (policy is not None) and (
```
Сильно шакально будет, если объединить 2 проверки на ```if policy```?

Misha-Shvets replied:
да

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
+            List of user group IDs, optional
+        :return: Select query
+        """
+        protocol_field = getattr(NetworkPolicy, protocol_type)
```
NetworkPolicy.protocol_type?

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
+                selectinload(qa(NetworkPolicy.groups)),
+                selectinload(qa(NetworkPolicy.mfa_groups)),
+            )
+            .filter(
```
У тебя 2 разных фильтра получается в запросе
1. filter_by
2. filter

Объедини

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
+            .limit(1)
+        )
+
+        if user_group_ids is not None:
```
Это точно правильная проверка? Если у тебя придет пустой список, то будет фильтр ```qa(Group.id).in_([])```

Misha-Shvets replied:
да

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
+        if user_group_ids is not None:
+            return query.filter(
+                or_(
+                    qa(NetworkPolicy.groups) == None,  # noqa
```
Дело говорит

Commit:
Refactor: update query filter for network policy group checks

- Changed the query filter in `NetworkPolicyGateway` to use `is_(None)` instead of `== None` for better clarity and adherence to SQLAlchemy best practices.

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
+            )
+            .filter(
+                text(':ip <<= ANY("Policies".netmasks)').bindparams(ip=ip),
+                protocol_field == true(),
```
Может тогда типа такого сделать ```text(f' "Policies"{protocol_type}  is  true')```
Но мне оба варианта не очень нравятся

Misha-Shvets replied:
ну и не будем трогать

Mastermind-U replied:
Bindparams гугли

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
         )
         return result.one()
+
+    def _build_base_query(
```
Я не вижу дальнейшего расширения результата этого метода. Возможно стоит переименовать, либо вынести фильтры, которые ты накидываешь под ```if user_group_ids ```, в  ```get_user_network_policy```

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
+                protocol_field == true(),
+            )
+            .order_by(qa(NetworkPolicy.priority).asc())
+            .limit(1)
```
Я фанат first'ов)

Misha-Shvets replied:
а я нет

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
+        :param NetworkPolicy policy: db policy
+        :return bool: status
+        """
+        if user is None or policy is None:
```
not (user and policy)

TheMihMih commented on app/ldap_protocol/policies/network/gateway.py:
```diff
@@ -134,3 +138,121 @@ async def check_policy_exists(self, policy: NetworkPolicy) -> bool:
+        )
+
+        group = await self._session.scalar(query)
+        return bool(group)
```
Может тогда делать exists запрос?  Да и можно объединить будет ```return await self._session.scalar(query)```

Misha-Shvets replied:
вкусовщина, как по мне

TheMihMih commented on app/ldap_protocol/policies/network/validator_gateway.py:
```diff
@@ -0,0 +1,77 @@
+from ldap_protocol.policies.network.gateway import NetworkPolicyGateway
+
+
+class NetworkPolicyValidatorGateway:
```
Весь этот класс выглядит как большая прокся. Он точно нужен? Нельзя использовать  NetworkPolicyGateway вместо этого

Misha-Shvets replied:
да

Mastermind-U replied:
Смысл был в том чтобы вынести сюда отдельно запросы для валидации сети. Простой гейтвей тут вообще не нужен, его тут не должно быть

TheMihMih commented on app/enums.py:
```diff
@@ -216,3 +216,11 @@ def combine(
+class ProtocolType(StrEnum):
+    """Protocol fields."""
+
+    LDAP = "is_ldap"
```
Кажется, что нейминг атрибута, совпадающий со значением был бы лучше, но тут значение как будто какое то булевое. Смотри сам крч

TheMihMih reviewed (changes requested):
Не хочешь написать тесты на блку NetworkPolicyValidatorGateway

Commit:
Refactor: streamline network policy imports and improve validation handling

- Consolidated import statements for `NetworkPolicyGateway`, `NetworkPolicyUseCase`, and `NetworkPolicyValidatorProtocol` across multiple files for better organization.
- Updated the `BindRequest` class to use the `policy` variable directly instead of accessing it through `ctx.ldap_session`.
- Enhanced the `NetworkPolicyGateway` class to simplify user and policy checks.
- Adjusted tests to align with the new import structure and validation methods.

Mastermind-U reviewed (approved):
Нужен отдельный гейтвей со своими sql запросами, текущий гейтвей надо разделить
Нужен отдельный юзкейс для унификации, пока что просто "прокси"

Misha-Shvets committed:
Http codes with map 774 (#836)

Commit:
Enhance network policy validation structure and imports

- Introduced new `NetworkPolicyValidatorUseCase` to streamline network policy validation processes.
- Updated `AuthorizationRules` to include new validation permissions for various network policies.
- Refactored `HTTPProvider` and `LDAPSession` to utilize the new validator use case instead of the deprecated gateway.
- Adjusted import statements across multiple files for better organization and clarity.
- Updated tests to align with the new validation structure and ensure proper functionality.

Commit:
Refactor: update network policy handling in HTTPProvider and LDAPSession

- Changed the scope of `network_policy_gateway`, `network_policy_validator`, and `network_policy_use_case` to `Scope.REQUEST` for improved request handling.
- Simplified the `LDAPSession` initialization by removing the `network_policy_validator_use_case` parameter.
- Enhanced the `validate_conn` method in `LDAPSession` to utilize `NetworkPolicyValidatorGateway` directly.
- Updated session management in `PoolClientHandler` and `CLDAPUDPServer` to pass the `AsyncSession` for network policy validation.
- Adjusted tests to align with the new session and validation structure.

Commit:
Fix: update parameter documentation in NetworkPolicyValidatorGateway

- Clarified the docstring for the `build_base_query` method by correcting the description of the `protocol_type` parameter to specify it as a filter for protocols.

TheMihMih commented on app/ldap_protocol/dialogue.py:
```diff
@@ -142,21 +146,17 @@ async def lock(self) -> AsyncIterator[UserSchema | None]:
     ) -> None:
         """Validate network policies."""
-        policy = await self._get_policy(ip, session)  # type: ignore
+        gateway = NetworkPolicyValidatorGateway(session)
```
Это надо передать в сам метод, если через ioc не получилось

TheMihMih commented on app/ioc.py:
```diff
@@ -650,18 +668,28 @@ def get_krb_template_render(
+        session: AsyncSession,
+    ) -> NetworkPolicyValidatorUseCase:
+        """Create network policy use case with request scope."""
+        gateway = NetworkPolicyValidatorGateway(session)
```
Так сделай 2 провайдера нормальных

Commit:
Refactor: enhance network policy validation integration

- Updated the `LDAPServerProvider` to streamline the provision of `NetworkPolicyValidatorUseCase` and `NetworkPolicyValidatorGateway`.
- Modified `validate_conn` method in `LDAPSession` to accept `NetworkPolicyValidatorUseCase` directly, improving clarity and reducing dependency on `AsyncSession`.
- Adjusted `PoolClientHandler` and `CLDAPUDPServer` to utilize the new validation structure, ensuring consistent network policy checks across components.

Commit:
Refactor: remove unused AsyncSession import in CLDAPUDPServer

- Eliminated the import of `AsyncSession` in `udp_server.py` as it is no longer needed following recent refactoring of network policy validation integration.
- Streamlined the code for better clarity and maintainability.

Commit:
Refactor: update network policy validation dependencies

- Changed the provision of `NetworkPolicyValidatorGateway` to `NetworkPolicyValidatorProtocol` in `HTTPProvider` for improved consistency in validation handling.
- Updated the constructor of `NetworkPolicyValidatorUseCase` to accept `NetworkPolicyValidatorProtocol`, aligning with recent refactoring efforts to streamline network policy validation processes.

Commit:
Merge branch 'dev' into refactor_network_policy_1015

Commit:
Refactor: update network policy validation provision in TestProvider

- Added provision for `NetworkPolicyValidatorProtocol` in `TestProvider` to enhance consistency in network policy validation.
- Adjusted the dependency injection for `NetworkPolicyValidatorGateway` to align with recent refactoring efforts.
